internal/provider/resource_tfe_workspace_agent_pool_execution.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceTFEWorkspaceAgentPoolExecution() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTFEWorkspaceAgentPoolExecutionCreate,
+		Read:   resourceTFEWorkspaceAgentPoolExecutionRead,
+		Update: resourceTFEWorkspaceAgentPoolExecutionUpdate,
+		Delete: resourceTFEWorkspaceAgentPoolExecutionDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"agent_pool_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"workspace_id": {
+				Type:     schema.TypeSet,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceTFEWorkspaceAgentPoolExecutionCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	poolID := d.Get("agent_pool_id").(string)
+	workspaceID := d.Get("workspace_id").(string)
+
+	// Create a new options struct to attach the agent pool to workspace
+	options := tfe.WorkspaceUpdateOptions{
+		AgentPoolID: tfe.String(poolID),
+	}
+
+	log.Printf("[DEBUG] Create attachment on workspace with agent pool ID: %s", poolID)
+	workspace, err := config.Client.Workspaces.UpdateByID(ctx, workspaceID, options)
+	if err != nil {
+		return fmt.Errorf("error attaching agent pool ID %s to workspace ID %s: %w", poolID, workspaceID, err)
+	}
+
+	d.SetId(workspace.ID)
+
+	return resourceTFEWorkspaceAgentPoolExecutionRead(d, meta)
+}
+
+func resourceTFEWorkspaceAgentPoolExecutionRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	log.Printf("[DEBUG] Read configuration: %s", d.Id())
+	workspace, err := config.Client.Workspaces.ReadByID(ctx, d.Id())
+	if err != nil {
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Workspace %s no longer exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error reading configuration of workspace %s: %w", d.Id(), err)
+	}
+
+	d.Set("workspace_id", workspace.ID)
+
+	var poolID string
+	if workspace.AgentPool != nil {
+		poolID = workspace.AgentPool.ID
+	}
+	d.Set("agent_pool_id", poolID)
+
+	return nil
+}
+
+func resourceTFEWorkspaceAgentPoolExecutionUpdate(d *schema.ResourceData, meta interface{}) error {
```

Comment:
My thought is that since the 2 params, i.e `agent_pool_id` and `workspace_id` are both `ForceNew: true`, this update function might not be needed because the resources will always be recreated as opposed to updated in-place.

Reply:
That's what the error message also noted in the CI, but I left it to have a discussion first. Seems like most of our resources have `Update` methods so I wondered what others think? Remove the `ForceNew: true` field or remove the `Update` method?

Reply:
Or, rather, most of our resources follow CRUD.